compiler/noirc_evaluator/src/ssa/opt/unrolling.rs

```diff
@@ -206,23 +270,22 @@ impl Loops {
     /// loop_end    loop_body
     /// ```
     /// `loop_entry` has two predecessors: `main` and `loop_body`, and it dominates `loop_body`.
+    ///
+    /// Returns all groups of blocks that look like a loop, even if we might not be able to unroll them,
+    /// which we can use to check whether we were able to unroll all blocks.
     pub(super) fn find_all(function: &Function) -> Self {
         let cfg = ControlFlowGraph::with_function(function);
         let post_order = PostOrder::with_function(function);
         let mut dom_tree = DominatorTree::with_cfg_and_post_order(&cfg, &post_order);
 
         let mut loops = vec![];
 
-        for (block, _) in function.dfg.basic_blocks_iter() {
-            // These reachable checks wouldn't be needed if we only iterated over reachable blocks
-            if dom_tree.is_reachable(block) {
-                for predecessor in cfg.predecessors(block) {
-                    // In the above example, we're looking for when `block` is `loop_entry` and `predecessor` is `loop_body`.
-                    if dom_tree.is_reachable(predecessor) && dom_tree.dominates(block, predecessor)
-                    {
-                        // predecessor -> block is the back-edge of a loop
-                        loops.push(Loop::find_blocks_in_loop(block, predecessor, &cfg));
-                    }
+        for block in function.reachable_blocks() {
```

Comment:
We could just use the reverse post order here. We already have the `post_order` in this function for computing the cfg

Reply:
For one it allows us to avoid going through all the blocks again. This would be a pretty minor optimization though. It is more that using `reachable_blocks` has bitten us before with hidden bugs (e.g., https://github.com/noir-lang/noir/pull/9495). It is usually only an issue with traversal order so I think we are safe here, but in general I think it would be good to move away from `reachable_blocks` in anything but testing code.

Reply:
I won't block on this though because looking at it again I think we are safe.